### PR TITLE
P2P shuffle skeleton

### PR DESCRIPTION
**Pull request by gjoseph92**
Somewhat implementing step one of the design in https://github.com/dask/distributed/pull/5435.

This skeleton is just intended to establish a framework for how the out-of-band shuffle will interact with dask (both with the workers, and with the task graph). The shuffle itself is intentionally not performant at all, but is correct.

Note that this does _not_ implement the scheduler extension in [mentioned](https://github.com/gjoseph92/distributed/blob/p2p-shuffle/proposal/distributed/shuffle/shuffle-design.md#scheduler-shuffle-extension) in the design doc yet. This requires a more substantial refactor than expected.

cc @fjetter 

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/shuffle/shuffle_extension.py [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+            task = asyncio.create_task(
+                self.worker.rpc(addr).shuffle_receive(
+                    shuffle_id=self.metadata.id,
+                    output_partition=output_partition,
+                    data=to_serialize(data),
+                )
+            )
+            tasks.append(task)
+
+        # TODO handle errors and cancellation here
```
Comment (fjetter) [inline distributed/shuffle/shuffle_extension.py:98]: Add a comment that we need the rerun group to do it properly which is why we haven't implemented anything, yet

File: distributed/shuffle/shuffle_extension.py [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+        self.transferred = False
+
+    def receive(self, output_partition: int, data: pd.DataFrame) -> None:
+        assert not self.transferred, "`receive` called after barrier task"
+        self.output_partitions[output_partition].append(data)
+
+    async def add_partition(self, data: pd.DataFrame) -> None:
+        assert not self.transferred, "`add_partition` called after barrier task"
+        tasks = []
+        # TODO grouping is blocking, should it be offloaded to a thread?
```
Comment (fjetter) [inline distributed/shuffle/shuffle_extension.py:85]: Don't think so, at least not now. Maybe once everything works. I suggest to remove the TODO

File: distributed/shuffle/shuffle_extension.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+        # It mostly doesn't release the GIL though, so may not make much difference.
+        for output_partition, data in data.groupby(self.metadata.column):
+            addr = self.metadata.worker_for(int(output_partition))
```
[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/shuffle/shuffle_extension.py:88]: I should have added a comment for this. The column will always be an integer column, which is the output partition number that each row belongs to. This is calculated using a hash within `shuffle`: https://github.com/dask/dask/blob/b196e4a500ce7d189cd2fad7bcf2864fa34b1e17/dask/dataframe/shuffle.py#L387-L404

Then that new column name is passed into `rearrange_by_column`. I find this interface a bit odd (and it's not documented anywhere in `rearrange_by_column` that the column has to be integer; in fact `rearrange_by_column_tasks` behaves differently), but we're just being consistent with that pattern.

File: distributed/shuffle/shuffle_extension.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+                f"Shuffle {new_metadata.id!r} is already registered on worker {self.worker.address}"
+            )
+
+        client = self.worker.client
+        assert (
+            client.loop is self.worker.loop
+        ), f"Worker client is not using the worker's event loop: {client.loop} vs {self.worker.loop}"
+        # NOTE: `Client.scheduler_info()` doesn't actually do anything when the client is async,
+        # manually call into the method for now.
+        await client._update_scheduler_info()
```
[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/shuffle/shuffle_extension.py:225]: See the implementation of `Client.scheduler_info`. It's a no-op if the client is asynchronous:
https://github.com/dask/distributed/blob/b56d15d2e4ed9b2c38d811cd360a0c71743e4aa4/distributed/client.py#L3511-L3513

Maybe I'm not understanding why. I know scheduler info is updated automatically every second:
https://github.com/dask/distributed/blob/b56d15d2e4ed9b2c38d811cd360a0c71743e4aa4/distributed/client.py#L721-L723
but I don't know why asking for the freshest possible info would be a no-op only with an async client.

File: distributed/shuffle/shuffle_extension.py [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+
+
+def add_handler(handlers: dict, handler: Callable) -> None:
+    "Add a handler to a worker's handlers dict, checking if it already exists."
+    existing_handler = handlers.setdefault(handler.__name__, handler)
+    if existing_handler is not handler:
+        raise RuntimeError(
+            f"Worker already has a {handler.__name__!r} handler registered "
+            f"from a different shuffle extension: {existing_handler}"
+        )
```
[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/shuffle/shuffle_extension.py:348]: It's not. I removed it in https://github.com/dask/distributed/pull/5524. This is overly defensive code from before I realized the extension would be baked into the worker at startup (as opposed to added by tasks/users at runtime).

File: distributed/shuffle/shuffle_extension.py [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+            )
+        i = len(self.workers) * output_partition // self.npartitions
+        return self.workers[i]
+
+    def partition_range(self, worker: str) -> tuple[int, int]:
+        "Get the output partition numbers (inclusive) that a worker will hold"
+        i = self.workers.index(worker)
+        first = math.ceil(self.npartitions * i / len(self.workers))
+        last = math.ceil(self.npartitions * (i + 1) / len(self.workers)) - 1
+        return first, last
```
[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/shuffle/shuffle_extension.py:60]: I'm not sure what you mean? Do you think this test of it is missing something?
https://github.com/dask/distributed/blob/0e403ec9364b720c8469ddda74199004fb601cbe/distributed/shuffle/tests/test_shuffle_extension.py#L52-L61

[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/shuffle/shuffle_extension.py:60]: I'm mostly wondering why a worker has a continuous range of IDs. I guess this connects to how we calculate the hash and partition ID. Can you put that information. Is it documented somewhere how all of these things connect and which ID is which and how they are calcualted, etc.

[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/shuffle/shuffle_extension.py:60]: > I'm mostly wondering why a worker has a continuous range of IDs.

That's an implementation detail of the binning arithmetic (`//` instead of `%`). I'll make this method private. It should not be required, nor relied on as part of the shuffle interface, that workers receive consecutive output partitions (though it is good for scheduling). This method just made it more readable and intuitive to implement `npartitions_for`, which _is_ needed.

File: distributed/shuffle/shuffle.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+def shuffle_barrier(id: ShuffleId, transfers: list[None]) -> None:
+    get_ext().barrier(id)
+
+
+def rearrange_by_column_p2p(
+    df: DataFrame,
+    column: str,
+    npartitions: int | None = None,
+):
```
Comment (fjetter) [inline distributed/shuffle/shuffle.py:55]:
```suggestion
) -> DataFrame:
```

File: distributed/shuffle/shuffle.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+            barrier_key, barrier_dsk, dependencies=[transferred]
+        ),
+    )
+
+    name = "shuffle-unpack-" + token
+    dsk = {
+        (name, i): (shuffle_unpack, token, i, barrier_key) for i in range(npartitions)
+    }
+    # TODO use this blockwise (https://github.com/coiled/oss-engineering/issues/49)
+    # Changes task names, so breaks setting worker restrictions at the moment.
```
Comment (fjetter) [inline distributed/shuffle/shuffle.py:91]: IIUC, neither of these PRs solves this problems, do they? Do we have an idea how to solve this problem down the road? What would need to change to enable this?

File: distributed/shuffle/shuffle_extension.py [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+
+        metadata = self._get_shuffle(shuffle_id).metadata
+
+        # Set worker restrictions for unpack tasks
+
+        # Could do this during `create_shuffle`, but we might as well overlap it with the time
+        # workers will be flushing buffers to each other.
+        name = "shuffle-unpack-" + metadata.id  # TODO single-source task name
+
+        # FIXME TODO XXX what about when culling means not all of the output tasks actually exist??!
```
Comment (fjetter) [inline distributed/shuffle/shuffle_extension.py:286]: I don't know what to do about the additional transfer but the cleanup/output task pruning could be handled here by accessing 
```python
worker = get_worker()
key = worker.get_current_task()
ts = worker.tasks[key]
# assert ts == barrier
ts.dependents
```

Comment (fjetter) [reply distributed/shuffle/shuffle_extension.py:286]: I assume knowing which dependents belongs to which ID boils down to the same problem we have on the scheduler plugin side, wouldn't it? The only way to properly assign the subset of dependents would be to parse they key for it's ID? Is there any other method to do this properly upfront during setup?

Comment (fjetter) [reply distributed/shuffle/shuffle_extension.py:286]: I don't have a very thorough understanding of the HLG and Layers classes but just looking over the API/signatures I would assume we were able to wrap the entire shuffle graph into a single layer and handle the culling there. That would allow us to mutate transfer and barrier tasks solving both problems.

```python
class Layer:
    def cull(...):
        ...
```

https://github.com/dask/dask/blob/ecddc8df5869679cef5280c869674cbf0d9f60d9/dask/highlevelgraph.py#L112-L156


Similarly, if we rather wanted to use resources, annotations or other attributes, the layer materialization would be an ideal place where we could set all of this, couldnt' we? We don't need any further infrastructure for this, do we?

**Review by fjetter — COMMENTED** [review]
I'm fine with the extension implementation but am wondering about the graph. let's discuss this and we can move on. Likely this all should be a follow up anyhow

File: distributed/shuffle/shuffle_extension.py (hunk 9 above)
Comment (gjoseph92) [reply distributed/shuffle/shuffle_extension.py:286]: > the cleanup/output task pruning could be handled here by accessing `worker.tasks[key].dependents`

Maybe I don't know worker state well enough, but I didn't think the scheduler sent workers the full list of dependents for a task. Don't workers only fill in dependents based on the dependencies of other tasks they've been told about? And since none of the dependents of the barrier can run until the barrier completes, the scheduler won't have sent those tasks to any workers, so I'd assume `dependents` is empty at this moment.

If this isn't true, and workers have a wider view of the graph than just the tasks they've been told to run, then that would change things a lot.

> The only way to properly assign the subset of dependents would be to parse they key for it's ID?

You need to know which _output partition number_ that task is creating. By definition for DataFrames, the output partition number is always the `i` part of the key `(name, i)`. So that's what we parse out of the key.

> Is there any other method to do this properly upfront during setup?

You could annotate each task with the output partition number, as mentioned.

You could maybe, very sketchily, assume something about the ordering/priorities of the keys instead of parsing them. Obviously just sorting by key names would work in most cases, because the keys all have the `i` component, but that's really no different from parsing. I _think_ maybe `enumerate(sorted(output_tasks, key=lambda ts: ts.priority))` would happen to give you the output partition number correctly simply because `dask.order` uses the ordering of key names to break ties—so again, you're still using the information in the key names.

Going fully mad, you could deserialize and inspect the task itself, traversing through all the layers of SubraphCallables until you found a call to `unpack`, and look at the `i` argument passed into it.

The bottom line is that in dask/dask, key names are not opaque. In fact, they're the place where this information (the partition number) is _defined_. So it's not surprising that most things come back to key names.

Comment (gjoseph92) [reply distributed/shuffle/shuffle_extension.py:286]: > wrap the entire shuffle graph into a single layer and handle the culling there

This is a reasonable idea, and I think how HLGs were intended work. There are already special layers for task-based shuffles; it makes sense there'd be one for this as well. And yeah, this would solve a lot of our problems around culling, annotations, etc. I avoided doing this initially because the whole thing is easily expressible just with Blockwise, and the Layer API is so ill-specified and a pain to work with.

The problem with this approach is that we'd lose fusion, because it wouldn't be a Blockwise layer anymore, so `optimize_blockwise` wouldn't know how to fuse it. And then that would expose us to root task overproduction for both the inputs to the shuffle and the outputs from it. We'd want to write our own optimization step for this layer type to fuse it to Blockwise layers (big pain), or extend `optimize_blockwise` to know about this layer (odd), or somehow make it a subclass of Blockwise (possibly odd and tricky, since Blockwise is very much just one logical layer, and this would be two Blockwises plus a materialized layer in between).

-------

Also, I hope this illustrates just how much we have to bend ourselves in knots because the system can't schedule linear chains of tasks properly (https://github.com/dask/distributed/issues/5555).

If we had STA (or any other solution to this problem), I don't think we'd even be considering this scheduler plugin approach. We'd just write a simple Layer here with a cull method, not worry about fusion, not worry about unstable task names or parsing keys, not worry about propagating annotations, and do all the necessary logic within tasks, with a couple RPCs to the scheduler to get peers and set restrictions/resources.

File: distributed/shuffle/shuffle.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    #     name,
+    #     "i",
+    #     token,
+    #     None,
+    #     BlockwiseDepDict({(i,): i for i in range(npartitions)}),
+    #     "i",
+    #     barrier_key,
+    #     None,
+    #     numblocks={},
+    # )
```
Comment (fjetter) [inline distributed/shuffle/shuffle.py:104]: In an earlier offline conversation, I mentioned that I am doubtful whether or not this should actually be considered a blockwise operation. My line of argumentation is mostly about the fact that the individual pieces are not independently calculable. This blockwise definition you propose itself also doesn't really look healthy in my opinion since we're encoding the `BlockwiseDepDict` to define the output partitions even though we're, in fact, not dealing with an IO layer here. It seems we're using an extremely complex machinery here even though it is not necessarily required.

The most important reason why we'd want to use this blockwise layer here is to enable downstream task fusion. imho, this should not require us to define this layer as blockwise but rather subsequent blockwise operations should care and be capable to fuse with with this. We may need to teach it this first and in the intermediate term we can use blockwise. My point is, I consider blockwise as rather the shortcut and not the as final solution.

Regardless of how the fusion is supposed to work, I am doubtful about whether or not this is even required. To my understanding fusion is extremely important for the input/transfer layer since it needs to be fused with whatever the input is. E.g. if we're starting off with a `read_parquet`, this should be fused with the `transfer` task. This is the important root task overproduction scenario since we do not read 100 files before we engage in any transfer.
However, on the output/unpack side, the situation is not as clear to me. At the very least it's not as urgent. Once we reach `unpack` all data is on-cluster. As long as we do not implement data spilling, the data is even in-memory. As long as all data is in memory, this _should_ be mostly irrelevant. Delaying this further down the road would be fine for me since I could think of other ways of dealing with this problem than forcing fusion (e.g. the worker plugin/extension could enforce a limit on how many tasks are allowed to be processed/be in memory at the same time).

**Review by fjetter — COMMENTED** [review]
I think most of what we're still discussing evolves around how we define the graph and how it can be further modified down the line (culling and optimizing/fusing). My current best feeling is that we should deal with this as a dedicated layer implementation which would be a nice follow up item, wouldn't it?

If you agree with this approach, I'm fine with merging

File: distributed/shuffle/shuffle.py (hunk 10 above)
Comment (gjoseph92) [reply distributed/shuffle/shuffle.py:104]: > As long as all data is in memory, this _should_ be mostly irrelevant. Delaying this further down the road would be fine for me since I could think of other ways of dealing with this problem than forcing fusion

That's fair. You're right that if the data is all in memory anyway, overproduction is irrelevant. And there are probably other hacks we can do to protect against it on the unpack side.

I think making a dedicated layer is the best approach here.

**Comment by gjoseph92** [comment]
Yeah, I think we should make a layer for this. And yes, it can be a follow-up task. So let's merge this one instead of https://github.com/dask/distributed/pull/5524.

However, DO NOT MERGE this until https://github.com/dask/dask/pull/8392 is merged! Otherwise, distributed CI will fail.

**Comment by gjoseph92** [comment]
This now only requires https://github.com/dask/dask/pull/8559 in order to merge.

**Comment by gjoseph92** [comment]
This is now unblocked.
